File: server/src/intTest/java/org/tctalent/server/repository/db/JobSpecificationIntTest.java
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getChatPost;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getJobChat;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSalesforceJobOpp;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedCandidate;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedCountry;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedJobChatUser;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedPartner;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedSavedList;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedUser;

import java.util.Collections;
import java.util.List;
import java.util.Set;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.tctalent.server.model.db.Candidate;
import org.tctalent.server.model.db.ChatPost;
import org.tctalent.server.model.db.Country;
import org.tctalent.server.model.db.JobChat;
import org.tctalent.server.model.db.JobChatType;
import org.tctalent.server.model.db.JobOpportunityStage;
import org.tctalent.server.model.db.PartnerImpl;
import org.tctalent.server.model.db.SalesforceJobOpp;
import org.tctalent.server.model.db.SavedList;
import org.tctalent.server.model.db.User;
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest;
import org.tctalent.server.request.job.SearchJobRequest;

public class JobSpecificationIntTest extends BaseDBIntegrationTest {

  @Autowired
  private SalesforceJobOppRepository repo;
  @Autowired
  private SavedListRepository savedListRepository;
  @Autowired
  private CountryRepository countryRepository;
  @Autowired
  private UserRepository userRepository;
  @Autowired
  private PartnerRepository partnerRepository;
  @Autowired
  private JobChatRepository jobChatRepository;
  @Autowired
  private JobChatUserRepository jcuRepository;
  @Autowired
  private ChatPostRepository chatPostRepository;
  @Autowired
  private CandidateRepository candidateRepo;
  private SalesforceJobOpp job;
  private SavedList savedList;
  private Country savedCountry;
  private User loggedInUser;
  private SearchJobRequest request;
  private Specification<SalesforceJobOpp> spec;

  @BeforeEach
  public void setup() {
    assertTrue(isContainerInitialised());

    PartnerImpl savedPartner = getSavedPartner(partnerRepository);
    loggedInUser = getSavedUser(userRepository);
    loggedInUser.setPartner(savedPartner);
    userRepository.save(loggedInUser);
    savedList = getSavedSavedList(savedListRepository);
    savedCountry = getSavedCountry(countryRepository);

    request = new SearchJobRequest();
  }

  @Test
  public void testKeywordWithNothingName() {
    job = getSalesforceJobOpp();
    repo.save(job);
    request.setKeyword("NOTHING");
    spec = JobSpecification.buildSearchQuery(request, null);
    List<SalesforceJobOpp> results = repo.findAll(spec);
    assertNotNull(results);
    assertTrue(results.isEmpty());
  }

  @Test
  public void testKeywordCaseInsensitive() {
    job = getSalesforceJobOpp();
    job.setSubmissionList(savedList);
    repo.save(job);

    request.setKeyword(job.getName().toUpperCase());
    spec = JobSpecification.buildSearchQuery(request, null);
    List<SalesforceJobOpp> results = repo.findAll(spec);
    assertNotNull(results);
    assertFalse(results.isEmpty());
    assertEquals(1, results.size());
    assertEquals(job.getId(), results.getFirst().getId());
  }

  @Test
  public void testStages() {
    job = getSalesforceJobOpp();
    job.setSubmissionList(savedList);
    job.setStage(JobOpportunityStage.cvPreparation);
    repo.save(job);

    request.setStages(Collections.singletonList(JobOpportunityStage.cvPreparation));
    spec = JobSpecification.buildSearchQuery(request, null);
    List<SalesforceJobOpp> results = repo.findAll(spec);
    assertNotNull(results);
    assertFalse(results.isEmpty());
    assertEquals(1, results.size());
    assertEquals(job.getId(), results.getFirst().getId());
  }

  @Test
  public void testStagesFail() {
    job = getSalesforceJobOpp();
    job.setSubmissionList(savedList);
    repo.save(job);

    request.setStages(List.of(JobOpportunityStage.noJobOffer));
    spec = JobSpecification.buildSearchQuery(request, null);
    List<SalesforceJobOpp> results = repo.findAll(spec);
    assertNotNull(results);
    assertTrue(results.isEmpty());
  }

  @Test
  public void testDestinations() {
    job = getSalesforceJobOpp();
    job.setSubmissionList(savedList);
    job.setCountry(savedCountry);
    repo.save(job);

    Long countryId = savedCountry.getId();
    request.setDestinationIds(Collections.singletonList(countryId));
    spec = JobSpecification.buildSearchQuery(request, null);
    List<SalesforceJobOpp> results = repo.findAll(spec);
    assertNotNull(results);
    assertFalse(results.isEmpty());
    assertEquals(1, results.size());
    assertEquals(job.getId(), results.getFirst().getId());
  }

  @Test
  public void testClosed() {
    job = getSalesforceJobOpp();
    job.setSubmissionList(savedList);
    job.setClosed(true);
    repo.save(job);

    SearchJobRequest request = new SearchJobRequest();
    request.setSfOppClosed(true);

    spec = JobSpecification.buildSearchQuery(request, null);
    List<SalesforceJobOpp> results = repo.findAll(spec);
    assertNotNull(results);
    assertFalse(results.isEmpty());
    assertEquals(1, results.size());
    assertEquals(job.getId(), results.getFirst().getId());
  }

  @Test
  public void testGetStarred() {
    job = getSalesforceJobOpp();
    job.setSubmissionList(savedList);
    job.setStarringUsers(Set.of(loggedInUser));
    repo.save(job);

    request.setStarred(true);

    spec = JobSpecification.buildSearchQuery(request, loggedInUser);
    List<SalesforceJobOpp> results = repo.findAll(spec);
    assertNotNull(results);
    assertFalse(results.isEmpty());
    assertEquals(1, results.size());
    assertEquals(job.getId(), results.getFirst().getId());
  }

  @Test
  public void testGetStarredLoggedInUserNull() {
    job = getSalesforceJobOpp();
    job.setStarringUsers(Set.of(loggedInUser));
    repo.save(job);

    request.setStarred(true);
    spec = JobSpecification.buildSearchQuery(request, null);
    List<SalesforceJobOpp> results = repo.findAll(spec);
    assertNotNull(results);
    assertTrue(results.isEmpty());
  }

  @Test
  public void testGetOwnedByMyPartner() {
    job = getSalesforceJobOpp();
    job.setSubmissionList(savedList);
    job.setCreatedBy(loggedInUser);
    repo.save(job);

    request.setSortFields(new String[]{"id"});
    request.setSortDirection(Sort.Direction.ASC);
    request.setOwnedByMyPartner(true);

    spec = JobSpecification.buildSearchQuery(request, loggedInUser);
    List<SalesforceJobOpp> results = repo.findAll(spec);
    assertNotNull(results);
    assertFalse(results.isEmpty());
    assertEquals(1, results.size());
    assertEquals(job.getId(), results.getFirst().getId());
  }

  @Test
  public void testGetOwnedByMeAndOrderByList() {
    job = getSalesforceJobOpp();
    job.setSubmissionList(savedList);
    job.setCreatedBy(loggedInUser);
    job.setContactUser(loggedInUser);
    repo.save(job);

    request.setSortFields(new String[]{"submissionList.id"});
    request.setSortDirection(Sort.Direction.ASC);
    request.setOwnedByMe(true);

    spec = JobSpecification.buildSearchQuery(request, loggedInUser);
    List<SalesforceJobOpp> results = repo.findAll(spec);
    assertNotNull(results);
    assertFalse(results.isEmpty());
    assertEquals(1, results.size());
    assertEquals(job.getId(), results.getFirst().getId());
  }

  @Test
  public void testGetOwnedByMeNoLoggedInUser() {
    job = getSalesforceJobOpp();
    job.setContactUser(loggedInUser);
    job.setCreatedBy(loggedInUser);
    repo.save(job);

    request.setOwnedByMe(true);
    spec = JobSpecification.buildSearchQuery(request, null);
    List<SalesforceJobOpp> results = repo.findAll(spec);
    assertNotNull(results);
    assertTrue(results.isEmpty());
  }

  @Test
  public void testUnreadMessages() {
    job = getSalesforceJobOpp();
    job.setSubmissionList(savedList);
    job.setCreatedBy(loggedInUser);
    job.setContactUser(loggedInUser);
    repo.save(job);

    User loggedInUser = getSavedUser(userRepository);
    Candidate c = getSavedCandidate(candidateRepo, loggedInUser);

    JobChat jc = getJobChat();
    jc.setCandidate(c);
    jc.setType(JobChatType.JobCreatorSourcePartner);
    jc.setJobOpp(job);
    jobChatRepository.save(jc);

    getSavedJobChatUser(jcuRepository, c.getUser(), jc);
    ChatPost cp = getChatPost();
    cp.setJobChat(jc);
    chatPostRepository.save(cp);

    request.setWithUnreadMessages(true);
    spec = JobSpecification.buildSearchQuery(request, loggedInUser);
    List<SalesforceJobOpp> results = repo.findAll(spec);
    assertNotNull(results);
    assertFalse(results.isEmpty());
    assertEquals(1, results.size());
    assertEquals(job.getId(), results.getFirst().getId());
  }
}

File: server/src/intTest/kotlin/org/tctalent/server/repository/db/JobSpecificationIntTest.kt
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db

import kotlin.test.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Sort
import org.tctalent.server.model.db.*
import org.tctalent.server.repository.db.integrationhelp.*
import org.tctalent.server.request.job.SearchJobRequest

class JobSpecificationIntTest : BaseDBIntegrationTest() {
  @Autowired private lateinit var repo: SalesforceJobOppRepository
  @Autowired private lateinit var savedListRepository: SavedListRepository
  @Autowired private lateinit var countryRepository: CountryRepository
  @Autowired private lateinit var userRepository: UserRepository
  @Autowired private lateinit var partnerRepository: PartnerRepository
  @Autowired private lateinit var jobChatRepository: JobChatRepository
  @Autowired private lateinit var jcuRepository: JobChatUserRepository
  @Autowired private lateinit var sfJobOpportunityRepository: SalesforceJobOppRepository
  @Autowired private lateinit var chatPostRepository: ChatPostRepository
  @Autowired private lateinit var candidateRepo: CandidateRepository
  private lateinit var job: SalesforceJobOpp
  private lateinit var savedList: SavedList
  private lateinit var savedCountry: Country
  private lateinit var loggedInUser: User

  @BeforeTest
  fun setup() {
    assertTrue { isContainerInitialized() }

    val savedPartner = getSavedPartner(partnerRepository)
    loggedInUser = userRepository.save(getUser().apply { partner = savedPartner })
    savedList = getSavedSavedList(savedListRepository)
    savedCountry = getSavedCountry(countryRepository)
  }

  @Test
  fun `test keyword with nothing name`() {
    job = getSalesforceJobOpp()
    repo.save(job)
    val request = SearchJobRequest().apply { keyword = "NOTHING" }
    val spec = JobSpecification.buildSearchQuery(request, null)
    val results = repo.findAll(spec)
    assertNotNull(results)
    assertTrue { results.isEmpty() }
  }

  @Test
  fun `test keyword case insensitive`() {
    job = repo.save(getSalesforceJobOpp().apply { submissionList = savedList })

    val request = SearchJobRequest().apply { keyword = job.name.uppercase() }
    val spec = JobSpecification.buildSearchQuery(request, null)
    val results = repo.findAll(spec)
    assertNotNull(results)
    assertTrue { results.isNotEmpty() }
    assertEquals(1, results.count())
    assertEquals(job.id, results.first().id)
  }

  @Test
  fun `test stages`() {
    job =
      repo.save(
        getSalesforceJobOpp().apply {
          stage = JobOpportunityStage.cvPreparation
          submissionList = savedList
        }
      )
    val request = SearchJobRequest().apply { stages = listOf(JobOpportunityStage.cvPreparation) }
    val spec = JobSpecification.buildSearchQuery(request, null)
    val results = repo.findAll(spec)
    assertNotNull(results)
    assertTrue { results.isNotEmpty() }
    assertEquals(1, results.count())
    assertEquals(job.id, results.first().id)
  }

  @Test
  fun `test stages fail`() {
    job = repo.save(getSalesforceJobOpp().apply { submissionList = savedList })
    val request = SearchJobRequest().apply { stages = listOf(JobOpportunityStage.noJobOffer) }
    val spec = JobSpecification.buildSearchQuery(request, null)
    val results = repo.findAll(spec)
    assertNotNull(results)
    assertTrue { results.isEmpty() }
  }

  @Test
  fun `test destinations`() {
    job =
      repo.save(
        getSalesforceJobOpp().apply {
          submissionList = savedList
          country = savedCountry
        }
      )

    val countryId = savedCountry.id
    val request = SearchJobRequest().apply { destinationIds = listOf(countryId) }
    val spec = JobSpecification.buildSearchQuery(request, null)
    val results = repo.findAll(spec)
    assertNotNull(results)
    assertTrue { results.isNotEmpty() }
    assertEquals(1, results.count())
    assertEquals(job.id, results.first().id)
  }

  //  @Test
  //  fun `test closed fail`() {
  //    job =
  //      repo.save(
  //        getSalesforceJobOpp().apply {
  //          isClosed = false
  //          submissionList = savedList
  //        }
  //      )
  //
  //    val request = SearchJobRequest().apply { sfOppClosed = true }
  //
  //    val spec = JobSpecification.buildSearchQuery(request, loggedInUser)
  //    val results = repo.findAll(spec)
  //    assertNotNull(results)
  //    assertTrue { results.isEmpty() }
  //  }

  @Test
  fun `test closed`() {
    job =
      repo.save(
        getSalesforceJobOpp().apply {
          submissionList = savedList
          isClosed = true
        }
      )

    val request = SearchJobRequest().apply { sfOppClosed = true }

    val spec = JobSpecification.buildSearchQuery(request, null)
    val results = repo.findAll(spec)
    assertNotNull(results)
    assertTrue { results.isNotEmpty() }
    assertEquals(1, results.count())
    assertEquals(job.id, results.first().id)
  }

  @Test
  fun `test get starred`() {
    job =
      repo.save(
        getSalesforceJobOpp().apply {
          submissionList = savedList
          starringUsers = setOf(loggedInUser)
        }
      )
    val request = SearchJobRequest().apply { starred = true }

    val spec = JobSpecification.buildSearchQuery(request, loggedInUser)
    val results = repo.findAll(spec)
    assertNotNull(results)
    assertTrue { results.isNotEmpty() }
    assertEquals(1, results.count())
    assertEquals(job.id, results.first().id)
  }

  @Test
  fun `test get starred logged in user null`() {
    job = repo.save(getSalesforceJobOpp().apply { starringUsers = setOf(loggedInUser) })

    val request = SearchJobRequest().apply { starred = true }
    val spec = JobSpecification.buildSearchQuery(request, null)
    val results = repo.findAll(spec)
    assertNotNull(results)
    assertTrue { results.isEmpty() }
  }

  @Test
  fun `test get owned by my partner`() {
    job =
      repo.save(
        getSalesforceJobOpp().apply {
          submissionList = savedList
          createdBy = loggedInUser
        }
      )

    val request =
      SearchJobRequest().apply {
        sortFields = arrayOf("id")
        sortDirection = Sort.Direction.ASC
        ownedByMyPartner = true
      }

    val spec = JobSpecification.buildSearchQuery(request, loggedInUser)
    val results = repo.findAll(spec)
    assertNotNull(results)
    assertTrue { results.isNotEmpty() }
    assertEquals(1, results.count())
    assertEquals(job.id, results.first().id)
  }

  @Test
  fun `test get owned by me and order by list`() {
    job =
      repo.save(
        getSalesforceJobOpp().apply {
          submissionList = savedList
          createdBy = loggedInUser
          contactUser = loggedInUser
        }
      )

    val request =
      SearchJobRequest().apply {
        sortFields = arrayOf("submissionList.id")
        sortDirection = Sort.Direction.ASC
        ownedByMe = true
      }
    val spec = JobSpecification.buildSearchQuery(request, loggedInUser)
    val results = repo.findAll(spec)
    assertNotNull(results)
    assertTrue { results.isNotEmpty() }
    assertEquals(1, results.count())
    assertEquals(job.id, results.first().id)
  }

  @Test
  fun `test get owned by me no logged in user`() {
    job =
      repo.save(
        getSalesforceJobOpp().apply {
          createdBy = loggedInUser
          contactUser = loggedInUser
        }
      )

    val request = SearchJobRequest().apply { ownedByMe = true }
    val spec = JobSpecification.buildSearchQuery(request, null)
    val results = repo.findAll(spec)
    assertNotNull(results)
    assertTrue { results.isEmpty() }
  }

  @Test
  fun `test unread messages`() {
    job =
      repo.save(
        getSalesforceJobOpp().apply {
          submissionList = savedList
          createdBy = loggedInUser
          contactUser = loggedInUser
        }
      )
    val loggedInUser = getSavedUser(userRepository)
    val c = getSavedCandidate(candidateRepo, loggedInUser)
    val jc =
      jobChatRepository.save(
        getJobChat().apply {
          candidate = c
          type = JobChatType.JobCreatorSourcePartner
          jobOpp = job
        }
      )
    getSavedJobChatUser(jcuRepository, c.user, jc)
    chatPostRepository.save(getChatPost().apply { jobChat = jc })

    //    candidateOpportunity = repo.save(getCandidateOpportunity().apply { candidate = c })

    val request = SearchJobRequest().apply { withUnreadMessages = true }
    val spec = JobSpecification.buildSearchQuery(request, loggedInUser)
    val results = repo.findAll(spec)
    assertNotNull(results)
    assertTrue { results.isNotEmpty() }
    assertEquals(1, results.count())
    assertEquals(job.id, results.first().id)
  }
}
